src/clib/pio_darray.c

```diff
@@ -147,88 +155,104 @@ int PIOc_write_darray_multi(int ncid, const int *varids, int ioid, int nvars,
-            if (!mpierr)
-                mpierr = MPI_Bcast(&ioid, 1, MPI_INT, ios->compmaster, ios->intercomm);
-            if (!mpierr)
-                mpierr = MPI_Bcast(&arraylen, 1, MPI_OFFSET, ios->compmaster, ios->intercomm);
-            if (!mpierr)
-                mpierr = MPI_Bcast(array, arraylen * iodesc->piotype_size, MPI_CHAR, ios->compmaster,
-                                   ios->intercomm);
-            if (!mpierr)
-                mpierr = MPI_Bcast(&frame_present, 1, MPI_CHAR, ios->compmaster, ios->intercomm);
-            if (!mpierr && frame_present)
-                mpierr = MPI_Bcast((void *)frame, nvars, MPI_INT, ios->compmaster, ios->intercomm);
-            if (!mpierr)
-                mpierr = MPI_Bcast(&fillvalue_present, 1, MPI_CHAR, ios->compmaster, ios->intercomm);
-            if (!mpierr && fillvalue_present)
-                mpierr = MPI_Bcast((void *)fillvalue, nvars * iodesc->piotype_size, MPI_CHAR,
-                                   ios->compmaster, ios->intercomm);
-            if (!mpierr)
-                mpierr = MPI_Bcast(&flushtodisk_int, 1, MPI_INT, ios->compmaster, ios->intercomm);
-            LOG((2, "PIOc_write_darray_multi file->pio_ncid = %d nvars = %d ioid = %d arraylen = %d "
-                 "frame_present = %d fillvalue_present = %d flushtodisk = %d", file->pio_ncid, nvars,
-                 ioid, arraylen, frame_present, fillvalue_present, flushtodisk));
+            amsg_frame = (int *)calloc(nvars, sizeof(int));
+        }
+        if(!fillvalue_present)
+        {
+            amsg_fillvalue = calloc(nvars * iodesc->piotype_size, sizeof(char ));
         }
 
-        /* Handle MPI errors. */
-        if ((mpierr2 = MPI_Bcast(&mpierr, 1, MPI_INT, ios->comproot, ios->my_comm)))
-            return check_mpi(file, mpierr2, __FILE__, __LINE__);
-        if (mpierr)
-            return check_mpi(file, mpierr, __FILE__, __LINE__);
+        PIO_SEND_ASYNC_MSG(ios, msg, &ierr,
+            ncid, nvars, nvars, varids, ioid, arraylen,
+            arraylen * iodesc->piotype_size, array, frame_present,
+            nvars,
+            (frame_present) ? frame : amsg_frame, fillvalue_present,
+            nvars * iodesc->piotype_size, amsg_fillvalue, flushtodisk_int);
+
+        if(!frame_present)
+        {
+            free(amsg_frame);
+        }
+        if(!fillvalue_present)
+        {
+            free(amsg_fillvalue);
+        }
 
         /* Share results known only on computation tasks with IO tasks. */
         if ((mpierr = MPI_Bcast(&fndims, 1, MPI_INT, ios->comproot, ios->my_comm)))
-            check_mpi(file, mpierr, __FILE__, __LINE__);
+            check_mpi(NULL, file, mpierr, __FILE__, __LINE__);
         LOG((3, "shared fndims = %d", fndims));
     }
 
     /* if the buffer is already in use in pnetcdf we need to flush first */
-    if (file->iotype == PIO_IOTYPE_PNETCDF && vdesc0->iobuf)
```

Comment:
OK, I wonder if this is a change we should grab. Why is using file better than vdesc0 here?